qiita_db/analysis.py

```diff
@@ -883,47 +896,45 @@ def _build_mapping_file(self, samples):
                      WHERE filepath_type = 'qiime_map'
                         AND artifact_id IN (SELECT *
                                             FROM qiita.find_artifact_roots(%s))
-                     ORDER BY filepath_id DESC"""
-            _id, fp = qdb.util.get_mountpoint('templates')[0]
-            to_concat = []
+                     ORDER BY filepath_id DESC LIMIT 1"""
+            _, fp = qdb.util.get_mountpoint('templates')[0]
 
-            for pid, samples in viewitems(samples):
-                if len(samples) != len(set(samples)):
-                    duplicates = find_duplicates(samples)
-                    raise qdb.exceptions.QiitaDBError(
-                        "Duplicate sample ids found: %s"
-                        % ', '.join(duplicates))
-                # Get the QIIME mapping file
+            all_ids = set()
+            to_concat = []
+            for pid, samps in viewitems(samples):
                 qdb.sql_connection.TRN.add(sql, [pid])
-                qiime_map_fp = \
-                    qdb.sql_connection.TRN.execute_fetchindex()[0][1]
+                qm_fp = qdb.sql_connection.TRN.execute_fetchindex()[0][1]
+
                 # Parse the mapping file
-                qiime_map = pd.read_csv(
-                    join(fp, qiime_map_fp), sep='\t', keep_default_na=False,
-                    na_values=['unknown'], index_col=False,
-                    converters=defaultdict(lambda: str))
-                qiime_map.set_index('#SampleID', inplace=True, drop=True)
-                qiime_map = qiime_map.loc[samples]
-
-                duplicates = all_sample_ids.intersection(qiime_map.index)
-                if duplicates or len(samples) != len(set(samples)):
-                    # Duplicate samples so raise error
-                    raise qdb.exceptions.QiitaDBError(
-                        "Duplicate sample ids found: %s"
-                        % ', '.join(duplicates))
-                all_sample_ids.update(qiime_map.index)
-                to_concat.append(qiime_map)
+                qm = pd.read_csv(join(fp, qm_fp), sep='\t',
+                                 keep_default_na=False, na_values=['unknown'],
+                                 index_col=False,
+                                 converters=defaultdict(lambda: str))
+
+                # if we are not going to merge the duplicated samples
+                # append the pid to the sample name
+                if duplicated_samples:
+                    samps = set(samps) - all_ids
+                    all_ids.update(samps)
+                else:
+                    qm['#SampleID'] = qm['#SampleID'] + ".%d" % pid
```

Comment:
Probably I remember this wrong - but I think that we agreed it was better to prefix the samples with the pid rather than suffix? @ElDeveloper do you remember what we agreed in here?